fastrand.go

```diff
@@ -20,65 +20,62 @@ import (
 // is the concatenation of an initial seed and a 128-bit counter. Each time
 // the entropy is hashed, the counter is incremented.
 type randReader struct {
-	entropy chan []byte
-}
-
-// Read fills b with random data. It always returns len(b), nil.
-func (r *randReader) Read(b []byte) (int, error) {
-	n := 0
-	for n < len(b) {
-		n += copy(b[n:], <-r.entropy)
-	}
-	return n, nil
-}
-
-// fillEntropy continuously fills r.entropy with new entropy.
-func (r *randReader) fillEntropy() {
-	// Create a hasher and fill it with 64 bytes of entropy. Technically only 16
-	// should be needed, but the underlying RNG may not be secure.
-	h, _ := blake2b.New256(nil)
-	_, err := io.CopyN(h, rand.Reader, 64)
-	if err != nil {
-		panic("fastrand: no entropy available")
-	}
-	seed := h.Sum(nil)
-
-	for {
-		for i := uint64(0); i < math.MaxUint64; i++ {
-			// Update the seed.
-			*(*uint64)(unsafe.Pointer(&seed[0])) = i
-
-			// Hash the seed.
-			h.Reset()
-			h.Write(seed)
-
-			// Send the entropy down the entropy channel.
-			r.entropy <- h.Sum(nil)
-		}
-
-		// Re-seed the hasher. Use the entropy that existed previously,
-		// protecting against a compromised RNG.
-		h.Reset()
-		h.Write(seed[:])
-		seed = h.Sum(seed[:0])
-	}
+	progress uint64
+	entropy  [32]byte
 }
 
 // Reader is a global, shared instance of a cryptographically strong pseudo-
 // random generator. It uses blake2b as its hashing function. Reader is safe
 // for concurrent use by multiple goroutines.
 var Reader io.Reader
 
+// init provides the initial entropy for the reader that will seed all numbers
+// coming out of fastrand.
 func init() {
-	r := &randReader{
-		entropy: make(chan []byte, 1000),
-	}
-	for i := 0; i < runtime.NumCPU(); i++ {
-		go r.fillEntropy()
+	r := &randReader{}
+	n, err := rand.Read(r.entropy[:])
+	if err != nil || n != len(r.entropy) {
+		panic("not enough entropy to fill fastrand reader at startup")
 	}
 	Reader = r
 }
 
+// Read fills b with random data. It always returns len(b), nil.
+func (r *randReader) Read(b []byte) (int, error) {
+	// Update the progress and use the result + the seed entropy to form a
+	// unique seed to hash.
+	progress := atomic.AddUint64(&r.progress, 1)
+
+	// If it seems like the progress is about to reset, increment the first
+	// entropy bytes. Without this check, the cycle time is 2^64 iterations, but
+	// with the check it is a secure 2^128 iterations.
+	if progress > 1<<63 {
+		atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
+		atomic.StoreUint64(&r.progress, 0)
```

Comment:
It triggers at 1 << 63, a lot lower than the max. You need to trigger it early because you don't know how many callers are using it simultaneously, and you must be certain that no caller ever ends up with the same entropy as a previous.

I think that the race condition is benign because the early 64bits are guaranteed to be unique per thread. The only risk is that you read a value that was set a while ago and it's paired with an early 64bits that have also been used with that value.

If we're going to bother rolling over on the 128 bits I guess we should stomp the race condition too.